File: tuan05/matranchiphi.py
def solve_tsp():
    # Ma tran chi phi de bai muc D
    c = [
        [0, 10, 9, 22, 23],
        [10, 0, 16, 19, 14],
        [9, 16, 0, 7, 28],
        [22, 19, 7, 0, 8],
        [23, 14, 28, 8, 0]
    ]
    n = len(c)

    # Tim canh re nhat toan ma tran (u != v)
    c_min = float('inf')
    for i in range(n):
        for j in range(n):
            if i != j and c[i][j] < c_min:
                c_min = c[i][j]

    # 1. Che do CO cat nhanh
    nodes_cut = 0
    min_cost_cut = float('inf')
    best_tour_cut = []
    tour = [0] * (n + 1)
    visited = [False] * n
    visited[0] = True

    def branch_and_bound(i, current_cost):
        nonlocal nodes_cut, min_cost_cut, best_tour_cut
        nodes_cut += 1

        for v in range(n):
            if not visited[v]:
                cost = current_cost + c[tour[i - 1]][v]
                # Can duoi: chi phi da di + so canh con lai * c_min
                lower_bound = cost + (n - i) * c_min

                if lower_bound < min_cost_cut:
                    tour[i] = v
                    visited[v] = True

                    if i == n - 1:
                        nodes_cut += 1  # Nut la khép chu trinh
                        total_cost = cost + c[v][tour[0]]
                        if total_cost < min_cost_cut:
                            min_cost_cut = total_cost
                            tour[n] = tour[0]
                            best_tour_cut = list(tour)
                    else:
                        branch_and_bound(i + 1, cost)

                    visited[v] = False

    branch_and_bound(1, 0)

    # 2. Che do KHONG cat nhanh
    nodes_no_cut = 0
    min_cost_no_cut = float('inf')
    visited_nc = [False] * n
    visited_nc[0] = True
    tour_nc = [0] * (n + 1)

    def backtrack_no_cut(i, current_cost):
        nonlocal nodes_no_cut, min_cost_no_cut
        nodes_no_cut += 1

        for v in range(n):
            if not visited_nc[v]:
                cost = current_cost + c[tour_nc[i - 1]][v]
                tour_nc[i] = v
                visited_nc[v] = True

                if i == n - 1:
                    nodes_no_cut += 1  # Nut la khép chu trinh
                    total_cost = cost + c[v][tour_nc[0]]
                    if total_cost < min_cost_no_cut:
                        min_cost_no_cut = total_cost
                else:
                    backtrack_no_cut(i + 1, cost)

                visited_nc[v] = False

    backtrack_no_cut(1, 0)

    print(f"Canh re nhat toan ma tran c_min: {c_min}")
    print(f"Chi phi hanh trinh toi uu: {min_cost_cut}")
    print(f"Hanh trinh toi uu: {' -> '.join(map(str, best_tour_cut))}")
    print(f"So nut khi CO cat nhanh: {nodes_cut}")
    print(f"So nut khi KHONG cat nhanh: {nodes_no_cut}")

if __name__ == "__main__":
    solve_tsp()

File: tuan05/matranchiphi.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

const int INF = 1e9;

int n = 5;
vector<vector<int>> c = {
    {0, 10, 9, 22, 23},
    {10, 0, 16, 19, 14},
    {9, 16, 0, 7, 28},
    {22, 19, 7, 0, 8},
    {23, 14, 28, 8, 0}
};

int c_min = INF;

// 1. Che do CO cat nhanh
int nodes_cut = 0;
int min_cost_cut = INF;
vector<int> tour_cut;
vector<int> best_tour_cut;
vector<bool> visited_cut;

void branch_and_bound(int i, int current_cost) {
    nodes_cut++;

    for (int v = 0; v < n; v++) {
        if (!visited_cut[v]) {
            int cost = current_cost + c[tour_cut[i - 1]][v];
            int lower_bound = cost + (n - i) * c_min;

            if (lower_bound < min_cost_cut) {
                tour_cut[i] = v;
                visited_cut[v] = true;

                if (i == n - 1) {
                    nodes_cut++;
                    int total_cost = cost + c[v][tour_cut[0]];
                    if (total_cost < min_cost_cut) {
                        min_cost_cut = total_cost;
                        tour_cut[n] = tour_cut[0];
                        best_tour_cut = tour_cut;
                    }
                } else {
                    branch_and_bound(i + 1, cost);
                }

                visited_cut[v] = false;
            }
        }
    }
}

// 2. Che do KHONG cat nhanh
int nodes_no_cut = 0;
int min_cost_no_cut = INF;
vector<int> tour_nc;
vector<bool> visited_nc;

void backtrack_no_cut(int i, int current_cost) {
    nodes_no_cut++;

    for (int v = 0; v < n; v++) {
        if (!visited_nc[v]) {
            int cost = current_cost + c[tour_nc[i - 1]][v];
            tour_nc[i] = v;
            visited_nc[v] = true;

            if (i == n - 1) {
                nodes_no_cut++;
                int total_cost = cost + c[v][tour_nc[0]];
                if (total_cost < min_cost_no_cut) {
                    min_cost_no_cut = total_cost;
                }
            } else {
                backtrack_no_cut(i + 1, cost);
            }

            visited_nc[v] = false;
        }
    }
}

int main() {
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            if (i != j && c[i][j] < c_min) {
                c_min = c[i][j];
            }
        }
    }

    tour_cut.assign(n + 1, 0);
    visited_cut.assign(n, false);
    visited_cut[0] = true;

    branch_and_bound(1, 0);

    tour_nc.assign(n + 1, 0);
    visited_nc.assign(n, false);
    visited_nc[0] = true;

    backtrack_no_cut(1, 0);

    cout << "Canh re nhat toan ma tran c_min: " << c_min << "\n";
    cout << "Chi phi hanh trinh toi uu: " << min_cost_cut << "\n";
    cout << "Hanh trinh toi uu: ";
    for (int i = 0; i <= n; i++) {
        cout << best_tour_cut[i] << (i == n ? "" : " -> ");
    }
    cout << "\n";
    cout << "So nut khi CO cat nhanh: " << nodes_cut << "\n";
    cout << "So nut khi KHONG cat nhanh: " << nodes_no_cut << "\n";

    return 0;
}